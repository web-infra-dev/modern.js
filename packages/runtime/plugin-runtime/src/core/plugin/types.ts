import type {
  RuntimePlugin as BaseRuntimePlugin,
  RuntimePluginExtends,
} from '@modern-js/plugin-v2';
import type { Hooks } from '@modern-js/plugin-v2/runtime';
import type { RuntimeContext } from '../context/runtime';
import type { Plugin } from './base';

export type RuntimeHooks = Hooks<RuntimeConfig, RuntimeContext>;

export type RuntimeExtends = Required<
  RuntimePluginExtends<RuntimeConfig, RuntimeContext, {}, {}>
>;

export type RuntimePluginFuture<Extends extends RuntimePluginExtends = {}> =
  BaseRuntimePlugin<RuntimeExtends & Extends>;
export interface RuntimeConfig {
  plugins?: (Plugin | RuntimePluginFuture)[];
}
