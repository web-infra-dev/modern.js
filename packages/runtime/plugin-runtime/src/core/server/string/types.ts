import type { ReactElement } from 'react';
import type { RenderLevel } from '../../constants';

export interface Collector {
  collect?: (comopnent: ReactElement) => ReactElement;
  effect: () => void | Promise<void>;
}

export type ChunkSet = {
  renderLevel: RenderLevel;
  ssrScripts: string;
  jsChunk: string;
  cssChunk: string;
};
