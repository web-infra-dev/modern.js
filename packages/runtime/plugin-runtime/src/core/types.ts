import type { BaseSSRServerContext } from '@modern-js/types';
import type { RenderLevel } from './constants';
import type { LoaderResult } from './loader/loaderManager';

declare global {
  interface Window {
    _SSR_DATA?: SSRContainer;
    _ROUTER_DATA?: RouterSSRData;
  }
}

export interface SSRData {
  loadersData?: Record<string, LoaderResult | undefined>;
  initialData?: Record<string, unknown>;
  storeState?: any;
  [props: string]: any;
}
export interface RouteData {
  [routeId: string]: any;
}
export interface RouterSSRData {
  loaderData: RouteData;
  errors: RouteData | null;
}

export type SSRMode = 'string' | 'stream';

export interface SSRContainer {
  renderLevel: RenderLevel;
  mode: SSRMode;
  data?: SSRData; // string ssr data
  context?: {
    request: {
      params: Record<string, any>;
      query: Record<string, string>;
      pathname: string;
      host: string;
      url: string;
      headers?: Record<string, string | undefined>;
    };
    reporter?: {
      sessionId?: string;
    };
  };
}

type BuildHtmlCb = (tempalte: string) => string;

export type SSRServerContext = Pick<
  BaseSSRServerContext,
  | 'redirection'
  | 'response'
  | 'nonce'
  | 'mode'
  | 'loaderContext'
  | 'reporter'
  | 'logger'
  | 'metrics'
  | 'routeManifest'
> & {
  request: BaseSSRServerContext['request'] & {
    baseUrl: string;
    userAgent: string;
    cookie: string;
    cookieMap: Record<string, string>;
    raw: Request;
  };
  htmlModifiers: BuildHtmlCb[];
  loaderFailureMode?: 'clientRender' | 'errorBoundary';
  onError?: (e: unknown) => void;
  onTiming?: (name: string, dur: number) => void;
  inlineScript?: boolean;
};
