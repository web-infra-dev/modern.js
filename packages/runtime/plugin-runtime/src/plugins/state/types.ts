import { createStore } from '@modern-js-reduck/store';
import { ReduxLoggerOptions } from 'redux-logger';

export type PluginOptions = Parameters<typeof createStore>[0] & {
  /**
   * If it's false, the logger will be disabled.
   * If it's a object, it means options will pass to createLogger function
   */
  logger?: false | ReduxLoggerOptions;
  // todo: archive effects params
  effects?: any;
  /**
   * Default: false
   * When it's true, will remove immer plugin
   */
  disableImmer?: boolean;
  // todo: archive devtools params
  devtools?: any;
};
