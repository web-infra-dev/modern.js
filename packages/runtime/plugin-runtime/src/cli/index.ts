import path from 'path';
import {
  PLUGIN_SCHEMAS,
  createRuntimeExportsUtils,
  cleanRequireCache,
} from '@modern-js/utils';
import type { CliPlugin } from '@modern-js/core';
import PluginState from '../state/cli';
import PluginSSR from '../ssr/cli';
import PluginRouter from '../router/cli';

export default (): CliPlugin => ({
  name: '@modern-js/runtime',
  post: [
    '@modern-js/plugin-router',
    '@modern-js/plugin-ssr',
    '@modern-js/plugin-state',
    '@modern-js/plugin-design-token',
  ],
  usePlugins: [PluginState(), PluginRouter(), PluginSSR()],
  setup: api => {
    let runtimeExportsUtils: ReturnType<typeof createRuntimeExportsUtils> =
      {} as any;

    return {
      config() {
        const dir = api.useAppContext().internalDirectory;
        runtimeExportsUtils = createRuntimeExportsUtils(dir, 'index');
        return {
          runtime: {},
          runtimeByEntries: {},
          source: {
            alias: {
              '@modern-js/runtime$': runtimeExportsUtils.getPath(),
              /**
               * twin.macro inserts styled-components into the code during the compilation process
               * But it will not be installed under the user project.
               * So need to add alias
               */
              'styled-components': require.resolve('styled-components', {
                paths: [require.resolve('@modern-js/runtime-core')],
              }),
            },
          },
        };
      },
      validateSchema() {
        return PLUGIN_SCHEMAS['@modern-js/runtime'];
      },
      addRuntimeExports() {
        const runtimePackage = path.resolve(__dirname, '../../../../');
        runtimeExportsUtils.addExport(`export * from '${runtimePackage}'`);
      },
      async beforeRestart() {
        cleanRequireCache([
          require.resolve('../state/cli'),
          require.resolve('../router/cli'),
          require.resolve('../ssr/cli'),
        ]);
      },
    };
  },
});
