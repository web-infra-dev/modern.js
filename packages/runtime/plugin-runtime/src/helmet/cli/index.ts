import type { AppTools, CliPlugin } from '@modern-js/app-tools';

const PLUGIN_IDENTIFIER = 'helmet';

export const helmetPlugin = (): CliPlugin<AppTools> => ({
  name: '@modern-js/plugin-helmet',

  required: ['@modern-js/runtime'],

  setup: api => {
    return {
      _internalRuntimePlugins({ entrypoint, plugins }) {
        const { metaName } = api.useAppContext();

        plugins.push({
          name: PLUGIN_IDENTIFIER,
          path: `@${metaName}/runtime/helmet`,
          config: {},
        });
        return { entrypoint, plugins };
      },
    };
  },
});

export default helmetPlugin;
