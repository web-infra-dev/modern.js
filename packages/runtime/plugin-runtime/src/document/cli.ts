import path from 'path';
import React from 'react';
import ReactDomServer from 'react-dom/server';
import { build } from 'esbuild';
import type { CliPlugin, UserConfig } from '@modern-js/core';
import { createDebugger, findExists } from '@modern-js/utils';
import { Entrypoint } from '@modern-js/types/cli';

import {
  DOCUMENT_SCRIPTS_PLACEHOLDER,
  DOCUMENT_FILE_NAME,
  DOCUMENT_META_PLACEHOLDER,
  DOCUMENT_NO_SCRIPTE_PLACEHOLDER,
  PLACEHOLDER_REPLACER_MAP,
  DOC_EXT,
  DOCUMENT_SSR_PLACEHOLDER,
  DOCUMENT_CHUNKSMAP_PLACEHOLDER,
  DOCUMENT_SSRDATASCRIPT_PLACEHOLDER,
  HTML_SEPARATOR,
} from './constants';
import { DocumentContext } from './DocumentContext';

const debug = createDebugger('html_genarate');

const getDocumenByEntryName = function (
  entrypoints: Entrypoint[],
  entryName: string,
): string | undefined {
  const entryDir = entrypoints.find(
    item => item.entryName === entryName,
  )?.absoluteEntryDir;

  const docFile = findExists(
    DOC_EXT.map(item => `${entryDir}${path.sep}${DOCUMENT_FILE_NAME}.${item}`),
  );

  return docFile || undefined;
};

export default (): CliPlugin => ({
  name: '@modern-js/document',
  pre: ['@modern-js/plugin-analyze'],
  setup: async api => {
    // 获取给 document 组件调用的参数
    function getDocParams(params: {
      config: UserConfig;
      entryName: string;
      templateParameters: Record<string, unknown>;
    }) {
      const { config, templateParameters, entryName } = params;
      // 保持兼容和足够的信息，分为：process, config, templateParams
      return {
        processEnv: process.env,
        config: {
          output: config.output,
        },
        entryName,
        templateParams: templateParameters,
      };
    }
    const documentEntry = (
      entryName: string,
      // config: HtmlPluginConfig,
      templateParameters: Record<string, unknown>,
    ) => {
      const { entrypoints, internalDirectory } = api.useAppContext();
      // 查找 entry 下 document 文件
      //  若 entry 下无 document.tsx 文件，则用 main 的兜底
      let documentFilePath = getDocumenByEntryName(entrypoints, entryName);
      if (!documentFilePath) {
        documentFilePath = getDocumenByEntryName(entrypoints, 'main');
      }
      // if no document file, do nothing as default
      if (!documentFilePath) {
        return null;
      }

      return async ({ htmlWebpackPlugin }: { [option: string]: any }) => {
        // 获取注入参数
        const documentParams = getDocParams({
          config: api.useConfigContext(),
          entryName,
          templateParameters,
        });
        const htmlOutputFile = path.join(
          internalDirectory,
          `./document/_${entryName}.html.js`,
        );
        // 将 document 文件转成 html string
        await build({
          entryPoints: [documentFilePath!],
          // write: false,
          outfile: htmlOutputFile,
          platform: 'node',
          target: 'es6',
          loader: {
            '.ts': 'ts',
            '.tsx': 'tsx',
          },
          bundle: true,
          plugins: [
            {
              name: 'make-all-packages-external',
              setup(build) {
                // https://github.com/evanw/esbuild/issues/619#issuecomment-751995294
                build.onResolve(
                  { filter: /^[^./]|^\.[^./]|^\.\.[^/]/ },
                  args => {
                    let external = true;
                    // FIXME: windows external entrypoint
                    if (args.kind === 'entry-point') {
                      external = false;
                    }
                    return {
                      path: args.path,
                      external,
                    };
                  },
                );
              },
            },
          ],
        });

        const Document = (await import(htmlOutputFile)).default;

        const HTMLElement = React.createElement(
          DocumentContext.Provider,
          { value: documentParams },
          React.createElement(Document, null),
        );
        const html = ReactDomServer.renderToString(HTMLElement);

        debug("entry %s's document jsx rendered html: %o", entryName, html);

        const scripts = [
          htmlWebpackPlugin.tags.headTags
            .filter((item: any) => item.tagName === 'script')
            .join(''),
          htmlWebpackPlugin.tags.bodyTags.toString(),
        ].join('');

        const metas = [
          templateParameters.meta,
          htmlWebpackPlugin.tags.headTags
            .filter((item: any) => item.tagName !== 'script')
            .join(''),
        ].join('');

        // 替换 html 占位符（因为 string 转成 jsx 比较麻烦。所以，使用占位符 + 替换的方式）
        return `<!DOCTYPE html>${html}`
          .replace(DOCUMENT_META_PLACEHOLDER, metas)
          .replace(DOCUMENT_SSR_PLACEHOLDER, HTML_SEPARATOR)
          .replace(DOCUMENT_SCRIPTS_PLACEHOLDER, scripts)
          .replace(
            DOCUMENT_NO_SCRIPTE_PLACEHOLDER,
            PLACEHOLDER_REPLACER_MAP[DOCUMENT_NO_SCRIPTE_PLACEHOLDER],
          )
          .replace(
            DOCUMENT_CHUNKSMAP_PLACEHOLDER,
            PLACEHOLDER_REPLACER_MAP[DOCUMENT_CHUNKSMAP_PLACEHOLDER],
          )
          .replace(
            DOCUMENT_SSRDATASCRIPT_PLACEHOLDER,
            PLACEHOLDER_REPLACER_MAP[DOCUMENT_SSRDATASCRIPT_PLACEHOLDER],
          );
      };
    };
    return {
      config: () => {
        return {
          tools: {
            htmlPlugin: (options, entry) => {
              // just for reuse the baseParames calculate by builder:
              // https://github.com/modern-js-dev/modern.js/blob/1abb452a87ae1adbcf8da47d62c05da39cbe4d69/packages/builder/builder-webpack-provider/src/plugins/html.ts#L69-L103
              const hackParameters: Record<string, unknown> =
                typeof options?.templateParameters === 'function'
                  ? options?.templateParameters(
                      {} as any,
                      {} as any,
                      {} as any,
                      {} as any,
                    )
                  : { ...options?.templateParameters };

              const templateContent = documentEntry(
                entry.entryName,
                // options,
                hackParameters,
              );

              const documentHtmlOptions = templateContent
                ? {
                    templateContent,
                    inject: false,
                  }
                : {};

              return {
                ...options,
                ...documentHtmlOptions,
              };
            },
          },
        };
      },
    };
  },
});
