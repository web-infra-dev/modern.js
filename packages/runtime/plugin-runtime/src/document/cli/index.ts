/* eslint-disable @typescript-eslint/no-require-imports */
import path from 'path';
import React from 'react';
import ReactDomServer from 'react-dom/server';
import { build } from 'esbuild';
import type {
  CliPlugin,
  AppTools,
  NormalizedConfig,
} from '@modern-js/app-tools';
import { createDebugger, findExists, fs } from '@modern-js/utils';
import { Entrypoint } from '@modern-js/types/cli';

import { DocumentContext } from '../DocumentContext';
import {
  DOCUMENT_SCRIPTS_PLACEHOLDER,
  DOCUMENT_LINKS_PLACEHOLDER,
  DOCUMENT_FILE_NAME,
  DOCUMENT_META_PLACEHOLDER,
  PLACEHOLDER_REPLACER_MAP,
  DOC_EXT,
  DOCUMENT_SSR_PLACEHOLDER,
  DOCUMENT_CHUNKSMAP_PLACEHOLDER,
  DOCUMENT_SSRDATASCRIPT_PLACEHOLDER,
  DOCUMENT_SCRIPT_PLACEHOLDER_START,
  DOCUMENT_SCRIPT_PLACEHOLDER_END,
  HTML_SEPARATOR,
  DOCUMENT_COMMENT_PLACEHOLDER_START,
  DOCUMENT_COMMENT_PLACEHOLDER_END,
  DOCUMENT_STYLE_PLACEHOLDER_START,
  DOCUMENT_STYLE_PLACEHOLDER_END,
  DOCUMENT_TITLE_PLACEHOLDER,
  TOP_PARTICALS_SEPARATOR,
  HEAD_PARTICALS_SEPARATOR,
  BODY_PARTICALS_SEPARATOR,
} from '../constants';

const debug = createDebugger('html_genarate');

// get the entry document file,
// if not exist, fallback to src/
export const getDocumenByEntryName = function (
  entrypoints: Entrypoint[],
  entryName: string,
  fallbackDir?: string,
): string | undefined {
  const entryDir = entrypoints.find(
    item => item.entryName === entryName,
  )?.absoluteEntryDir;

  const entryDirs = DOC_EXT.map(
    item => `${entryDir}${path.sep}${DOCUMENT_FILE_NAME}.${item}`,
  );
  const fallbackDirs = fallbackDir
    ? DOC_EXT.map(item =>
        [fallbackDir, 'src', `${DOCUMENT_FILE_NAME}.${item}`].join(path.sep),
      )
    : [];

  const docFile = findExists([...entryDirs, ...fallbackDirs]);

  return docFile || undefined;
};

export const documentPlugin = (): CliPlugin<AppTools> => ({
  name: '@modern-js/plugin-document',

  pre: ['@modern-js/plugin-analyze'],
  setup: async api => {
    // get params for document.tsx
    function getDocParams(params: {
      config: NormalizedConfig<AppTools>;
      entryName: string;
      templateParameters: Record<string, unknown>;
    }) {
      const { config, templateParameters, entryName } = params;
      // for enough params, devide asï¼šprocess, config, templateParams
      return {
        processEnv: process.env,
        config: {
          output: config.output,
        },
        entryName,
        templateParams: templateParameters,
      };
    }
    const documentEntry = (
      entryName: string,
      // config: HtmlPluginConfig,
      templateParameters: Record<string, unknown>,
    ) => {
      const { entrypoints, internalDirectory, appDirectory } =
        api.useAppContext();
      // search the document.[tsx|jsx|js|ts] under entry
      const documentFilePath = getDocumenByEntryName(
        entrypoints,
        entryName,
        appDirectory,
      );
      // if no document file, do nothing as default
      if (!documentFilePath) {
        return null;
      }

      return async ({ htmlWebpackPlugin }: { [option: string]: any }) => {
        const config = api.useResolvedConfigContext();

        const documentParams = getDocParams({
          config,
          entryName,
          templateParameters,
        });

        // set a temporary tsconfig file for divide the influence by project's jsx
        const tempTsConfigFile = path.join(
          internalDirectory,
          `./document/_tempTsconfig.json`,
        );
        const userTsConfigFilePath = path.join(appDirectory, 'tsconfig.json');
        let tsConfig;
        try {
          // eslint-disable-next-line import/no-dynamic-require
          tsConfig = await require(userTsConfigFilePath);
        } catch (err) {
          tsConfig = {};
        }
        if (tsConfig?.compilerOptions) {
          tsConfig.compilerOptions.jsx = 'react-jsx';
        } else {
          tsConfig.compilerOptions = {
            jsx: 'react-jsx',
          };
        }
        fs.outputFileSync(tempTsConfigFile, JSON.stringify(tsConfig));

        const htmlOutputFile = path.join(
          internalDirectory,
          `./document/_${entryName}.html.js`,
        );
        // transform document file to html string
        await build({
          entryPoints: [documentFilePath],
          outfile: htmlOutputFile,
          platform: 'node',
          // change esbuild use the rootDir tsconfig.json as default to tempTsConfigFile
          tsconfig: tempTsConfigFile,
          target: 'es6',
          loader: {
            '.ts': 'ts',
            '.tsx': 'tsx',
          },
          bundle: true,
          plugins: [
            {
              name: 'make-all-packages-external',
              setup(build) {
                // https://github.com/evanw/esbuild/issues/619#issuecomment-751995294
                build.onResolve(
                  { filter: /^[^./]|^\.[^./]|^\.\.[^/]/ },
                  args => {
                    let external = true;
                    // FIXME: windows external entrypoint
                    if (args.kind === 'entry-point') {
                      external = false;
                    }
                    return {
                      path: args.path,
                      external,
                    };
                  },
                );
              },
            },
          ],
        });

        delete require.cache[require.resolve(htmlOutputFile)];
        // eslint-disable-next-line import/no-dynamic-require
        const Document = (await require(htmlOutputFile)).default;
        const HTMLElement = React.createElement(
          DocumentContext.Provider,
          { value: documentParams },
          React.createElement(Document, null),
        );
        let html = ReactDomServer.renderToStaticMarkup(HTMLElement);

        debug("entry %s's document jsx rendered html: %o", entryName, html);
        // htmlWebpackPlugin.tags
        const { partialsByEntrypoint } = api.useAppContext();
        const scripts = [
          htmlWebpackPlugin.tags.headTags
            .filter((item: any) => item.tagName === 'script')
            .join(''),
          htmlWebpackPlugin.tags.bodyTags.toString(),
        ].join('');
        // support partials html
        const partialsContent = {
          partialsTop: '',
          partialsHead: '',
          partialsBody: '',
        };
        if (partialsByEntrypoint?.[entryName]) {
          partialsContent.partialsTop =
            partialsByEntrypoint[entryName].top.join('\n');
          partialsContent.partialsHead =
            partialsByEntrypoint[entryName].head.join('\n');
          partialsContent.partialsBody =
            partialsByEntrypoint[entryName].body.join('\n');
        }

        html = html
          .replace(TOP_PARTICALS_SEPARATOR, partialsContent.partialsTop)
          .replace(HEAD_PARTICALS_SEPARATOR, partialsContent.partialsHead)
          .replace(BODY_PARTICALS_SEPARATOR, partialsContent.partialsBody);

        const links = [
          htmlWebpackPlugin.tags.headTags
            .filter((item: any) => item.tagName === 'link')
            .join(''),
        ].join('');

        const metas = [
          templateParameters.meta,
          htmlWebpackPlugin.tags.headTags
            .filter(
              (item: any) =>
                item.tagName !== 'script' && item.tagName !== 'link',
            )
            .join(''),
        ].join('');

        const titles = [
          templateParameters.title,
          htmlWebpackPlugin.tags.headTags
            .filter((item: any) => item.tagName === 'title')
            .join(''),
        ].join('');

        // if the Document.tsx has a functional script, replace to convert it
        if (
          html.includes(DOCUMENT_SCRIPT_PLACEHOLDER_START) &&
          html.includes(DOCUMENT_SCRIPT_PLACEHOLDER_END)
        ) {
          const { nonce } = config.security;
          const nonceAttr = nonce ? `nonce=${nonce}` : '';

          html = html.replace(
            new RegExp(
              `${DOCUMENT_SCRIPT_PLACEHOLDER_START}(.*?)${DOCUMENT_SCRIPT_PLACEHOLDER_END}`,
              'g',
            ),
            (_scriptStr, $1) =>
              `<script ${nonceAttr}>${decodeURIComponent($1)}</script>`,
          );
        }
        // if the Document.tsx has a style, replace to convert it
        if (
          html.includes(DOCUMENT_STYLE_PLACEHOLDER_START) &&
          html.includes(DOCUMENT_STYLE_PLACEHOLDER_END)
        ) {
          html = html.replace(
            new RegExp(
              `${DOCUMENT_STYLE_PLACEHOLDER_START}(.*?)${DOCUMENT_STYLE_PLACEHOLDER_END}`,
              'g',
            ),
            (_styleStr, $1) => `<style>${decodeURIComponent($1)}</style>`,
          );
        }
        // if the Document.tsx has a comment component, replace and convert it
        if (
          html.includes(DOCUMENT_COMMENT_PLACEHOLDER_START) &&
          html.includes(DOCUMENT_COMMENT_PLACEHOLDER_END)
        ) {
          html = html.replace(
            new RegExp(
              `${DOCUMENT_COMMENT_PLACEHOLDER_START}(.*?)${DOCUMENT_COMMENT_PLACEHOLDER_END}`,
              'g',
            ),
            (_scriptStr, $1) => `${decodeURIComponent($1)}`,
          );
        }

        // replace the html placeholder while transfer string to jsx component is not a easy way
        const finalHtml = `<!DOCTYPE html>${html}`
          .replace(DOCUMENT_META_PLACEHOLDER, metas)
          .replace(DOCUMENT_SSR_PLACEHOLDER, HTML_SEPARATOR)
          .replace(DOCUMENT_SCRIPTS_PLACEHOLDER, scripts)
          .replace(DOCUMENT_LINKS_PLACEHOLDER, links)
          .replace(
            DOCUMENT_CHUNKSMAP_PLACEHOLDER,
            PLACEHOLDER_REPLACER_MAP[DOCUMENT_CHUNKSMAP_PLACEHOLDER],
          )
          .replace(
            DOCUMENT_SSRDATASCRIPT_PLACEHOLDER,
            PLACEHOLDER_REPLACER_MAP[DOCUMENT_SSRDATASCRIPT_PLACEHOLDER],
          )
          .replace(DOCUMENT_TITLE_PLACEHOLDER, titles);
        return finalHtml;
      };
    };
    return {
      config: () => {
        return {
          tools: {
            htmlPlugin: (options, entry) => {
              // just for reuse the baseParames calculate by builder:
              // https://github.com/web-infra-dev/modern.js/blob/1abb452a87ae1adbcf8da47d62c05da39cbe4d69/packages/builder/builder-webpack-provider/src/plugins/html.ts#L69-L103
              const hackParameters: Record<string, unknown> =
                typeof options?.templateParameters === 'function'
                  ? options?.templateParameters(
                      {} as any,
                      {} as any,
                      {} as any,
                      {} as any,
                    )
                  : { ...options?.templateParameters };

              const templateContent = documentEntry(
                entry.entryName,
                // options,
                hackParameters,
              );

              const documentHtmlOptions = templateContent
                ? {
                    templateContent,
                    // Note: the behavior of inject/modify tags in afterTemplateExecution hook will not take effect
                    inject: false,
                  }
                : {};

              return {
                ...options,
                ...documentHtmlOptions,
              };
            },
          },
        };
      },
    };
  },
});

export default documentPlugin;
