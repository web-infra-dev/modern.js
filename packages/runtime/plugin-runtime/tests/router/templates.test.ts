/* eslint-disable @typescript-eslint/no-empty-function */
import { RouteLegacy } from '@modern-js/types/cli';
import {
  fileSystemRoutes,
  routesForServer,
} from '../../src/router/cli/code/templates';

jest.mock('@modern-js/utils', () => {
  const fs = {
    writeFile() {},
    writeJSON() {},
    ensureFile() {},
  };
  return {
    __esModule: true,
    ...jest.requireActual('@modern-js/utils'),
    fs,
  };
});

expect.addSnapshotSerializer({
  test: val => typeof val === 'string',
  print: (val: unknown) => (val as string).replace(/\\/g, '/'),
});

describe('fileSystemRoutes', () => {
  test('generate code for legacy router', async () => {
    const routes: RouteLegacy[] = [
      {
        path: '/user',
        exact: true,
        component: '@/pages/user',
        _component: '@/pages/user',
      },
    ];

    const code = await fileSystemRoutes({
      routes,
      entryName: 'main',
      internalDirectory: '',
    });
    expect(code).toMatchSnapshot();
  });

  test('generate code', async () => {
    const routes = [
      {
        path: '/',
        _component: '@_modern_js_src/routes/layout.tsx',
        id: 'layout',
        type: 'nested' as const,
        children: [
          {
            path: 'user',
            error: '@_modern_js_src/routes/error.tsx',
            _component: '@_modern_js_src/routes/user/layout.tsx',
            loading: '@_modern_js_src/routes/loading.tsx',
            id: 'user/layout',
            type: 'nested' as const,
            loader: '@_modern_js_src/routes/layout.loader.ts',
            children: [
              {
                path: ':id',
                id: 'user/[id]/layout',
                type: 'nested' as const,
                children: [
                  {
                    _component: '@_modern_js_src/routes/user/[id]/page.tsx',
                    index: true,
                    id: 'user/[id]/page',
                    loader: '@_modern_js_src/routes/user/[id]/page.tsx',
                    type: 'nested' as const,
                  },
                ],
              },
            ],
          },
        ],
      },
    ];
    const code = await fileSystemRoutes({
      entryName: 'main',
      routes,
      internalDirectory: '',
    });
    expect(code).toMatchSnapshot();
  });
});

describe('routesForServer', () => {
  test('generate code for server', async () => {
    const routes = [
      {
        path: '/',
        _component: '@_modern_js_src/routes/layout.tsx',
        id: 'layout',
        type: 'nested' as const,
        children: [
          {
            path: 'user',
            error: '@_modern_js_src/routes/error.tsx',
            _component: '@_modern_js_src/routes/user/layout.tsx',
            loading: '@_modern_js_src/routes/loading.tsx',
            id: 'user/layout',
            type: 'nested' as const,
            loader: '@_modern_js_src/routes/layout.loader.ts',
            children: [
              {
                path: ':id',
                id: 'user/[id]/layout',
                type: 'nested' as const,
                children: [
                  {
                    _component: '@_modern_js_src/routes/user/[id]/page.tsx',
                    index: true,
                    id: 'user/[id]/page',
                    loader: '@_modern_js_src/routes/user/[id]/page.loader.ts',
                    type: 'nested' as const,
                  },
                ],
              },
            ],
          },
        ],
      },
    ];
    const code = routesForServer({
      routes,
    });
    expect(code).toMatchSnapshot();
  });
});
