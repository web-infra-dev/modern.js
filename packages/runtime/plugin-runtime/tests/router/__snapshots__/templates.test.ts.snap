// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`fileSystemRoutes generate code 1`] = `

    
    import { lazy } from "react";
    import loadable, { lazy as loadableLazy } from "@modern-js/runtime/loadable"
  
    
    
    import { createShouldRevalidate, handleRouteModule,  handleRouteModuleError} from '@modern-js/runtime/router';
  
    
    import loading_0 from '@_modern_js_src/routes/loading.tsx';

    import error_0 from '@_modern_js_src/routes/error.tsx';

    import loader_0 from "@_modern_js_src/routes/user/[id]/page.tsx";
import loader_1 from "@_modern_js_src/routes/layout.loader.ts";

    
    
    if(typeof document !== 'undefined'){
      window._routeModules = {}
    }
  
    
    export const routes = [
  {
  "path": "/",
  "_component": "@_modern_js_src/routes/layout.tsx",
  "id": "layout",
  "type": "nested",
  "children": [
    {
      "path": "user",
      "error": error_0,
      "_component": "@_modern_js_src/routes/user/layout.tsx",
      "loading": loading_0,
      "id": "user/layout",
      "type": "nested",
      "loader": loader_1,
      "children": [
        {
          "path": ":id",
          "id": "user/[id]/layout",
          "type": "nested",
          "children": [
            {
              "_component": "@_modern_js_src/routes/user/[id]/page.tsx",
              "index": true,
              "id": "user/[id]/page",
              "loader": loader_0,
              "type": "nested",
              "lazyImport": () => import(/* webpackChunkName: "user/[id]/page" */  '@_modern_js_src/routes/user/[id]/page.tsx').then(routeModule => handleRouteModule(routeModule, "user/[id]/page")).catch(handleRouteModuleError) ,
              "component": lazy(() => import(/* webpackChunkName: "user/[id]/page" */  '@_modern_js_src/routes/user/[id]/page.tsx').then(routeModule => handleRouteModule(routeModule, "user/[id]/page")).catch(handleRouteModuleError) ),
              "shouldRevalidate": createShouldRevalidate("user/[id]/page")
            }
          ],
          "lazyImport": null
        }
      ],
      "lazyImport": () => import(/* webpackChunkName: "user/layout" */  '@_modern_js_src/routes/user/layout.tsx').then(routeModule => handleRouteModule(routeModule, "user/layout")).catch(handleRouteModuleError) ,
      "component": lazy(() => import(/* webpackChunkName: "user/layout" */  '@_modern_js_src/routes/user/layout.tsx').then(routeModule => handleRouteModule(routeModule, "user/layout")).catch(handleRouteModuleError) ),
      "shouldRevalidate": createShouldRevalidate("user/layout")
    }
  ],
  "lazyImport": () => import(/* webpackChunkName: "layout" */  '@_modern_js_src/routes/layout.tsx').then(routeModule => handleRouteModule(routeModule, "layout")).catch(handleRouteModuleError) ,
  "component": lazy(() => import(/* webpackChunkName: "layout" */  '@_modern_js_src/routes/layout.tsx').then(routeModule => handleRouteModule(routeModule, "layout")).catch(handleRouteModuleError) )
},
];
  
`;

exports[`fileSystemRoutes generate code for legacy router 1`] = `

    
    import { lazy } from "react";
    import loadable, { lazy as loadableLazy } from "@modern-js/runtime/loadable"
  
    
    
    import { createShouldRevalidate, handleRouteModule,  handleRouteModuleError} from '@modern-js/runtime/router';
  
    
    
    
    
    
    
    if(typeof document !== 'undefined'){
      window._routeModules = {}
    }
  
    
    export const routes = [
  {
  "path": "/user",
  "exact": true,
  "component": loadable(() => import('@/pages/user')),
  "_component": "@/pages/user"
},
];
  
`;

exports[`routesForServer generate code for server 1`] = `

    import loader_0 from "@_modern_js_src/routes/user/[id]/page.loader.ts";
import loader_1 from "@_modern_js_src/routes/layout.loader.ts";

    
  export const routes = [
  {
  "path": "/",
  "_component": "@_modern_js_src/routes/layout.tsx",
  "id": "layout",
  "type": "nested",
  "children": [
    {
      "path": "user",
      "error": "@_modern_js_src/routes/error.tsx",
      "_component": "@_modern_js_src/routes/user/layout.tsx",
      "loading": "@_modern_js_src/routes/loading.tsx",
      "id": "user/layout",
      "type": "nested",
      "loader": loader_1,
      "children": [
        {
          "path": ":id",
          "id": "user/[id]/layout",
          "type": "nested",
          "children": [
            {
              "_component": "@_modern_js_src/routes/user/[id]/page.tsx",
              "index": true,
              "id": "user/[id]/page",
              "loader": loader_0,
              "type": "nested"
            }
          ]
        }
      ]
    }
  ]
}
];
  
`;
